notifications-worker.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/dxe/alc-mobile-api/model"
+	"github.com/jmoiron/sqlx"
+	expo "github.com/oliveroneill/exponent-server-sdk-golang/sdk"
+)
+
+func filterValidTokens(notifications []model.Notification) ([]model.Notification, []expo.PushMessage) {
+	var validNotifications []model.Notification
+	var messages []expo.PushMessage
+	for _, n := range notifications {
+		pushToken, err := expo.NewExponentPushToken(n.ExpoPushToken)
+		if err != nil {
+			continue
+		}
+		validNotifications = append(validNotifications, n)
+		messages = append(messages, expo.PushMessage{
+			To:    []expo.ExponentPushToken{pushToken},
+			Title: n.Title,
+			Body:  n.Body,
+		})
+	}
+	return validNotifications, messages
+}
+
+const (
+	StatusSent                = "Sent"
+	StatusDeviceNotRegistered = "DeviceNotRegistered"
+	StatusUnknownError        = "UnknownError"
+)
+
+func notificationsWorker(db *sqlx.DB, client *expo.PushClient) (err error) {
+	notifications, err := model.SelectNotificationsToSend(db)
+	if err != nil {
+		return err
+	}
+
+	if len(notifications) == 0 {
+		return nil
+	}
+
+	validNotifications, validExpoMessages := filterValidTokens(notifications)
+
+	// TODO: use context to cancel if this takes too long?
+	expoResponses, err := client.PublishMultiple(validExpoMessages)
+	if err != nil {
+		return fmt.Errorf("failed to publish messages via expo api: %w", err)
+	}
+
+	// Update each notification with the status.
+	for i, r := range expoResponses {
+		if r.Status == expo.SuccessStatus {
+			validNotifications[i].Status = StatusSent
+			continue
```

Comment:
As-is is fine; but instead of `continue`, I'd probably structure this as:

```
switch {
case r.Status == expo.SuccessStatus:
    ...
case r.Details["error"] == expo.ErrorDeviceNotRegistered:
    ...
default:
    ...
}
```

I think this helps emphasize that there's a multiway branch happening here.

You can also do `if/else if`, but I personally tend to prefer `switch` over `else if`.

Reply:
Making this change.